File: backend/src/backend/schemas/ui.py
"""UI Data API (BFF) 응답 스키마.

Agent 는 `component` SSE 시그널만 보내고(ADR-002), FE 가 이 엔드포인트로 카드 데이터를
조회한다. 현재는 목 픽스처, 향후 postgres/redis/mock-financial-service 조회로 교체.
스키마는 backend/docs/agent_ui_event_spec.md §4b 와 일치한다.
"""

from pydantic import BaseModel, Field


class AccountSummary(BaseModel):
    id: int
    bank: str
    alias: str
    tail: str
    balance: int
    color: str


class BalanceData(BaseModel):
    """GET /api/v1/ui/balance"""

    total: int = Field(description="총 자산(원)")
    accounts: list[AccountSummary]


# --- account_detail (계좌 상세, GET /api/v1/ui/account/{account_id}) ---


class RecentTxItem(BaseModel):
    name: str
    emoji: str
    date: str
    amount: int = Field(description="부호 있음(출금 음수) — A2 규칙과 다름")
    type: str = Field(description="'in' | 'out'")


class AccountDetailInfo(BaseModel):
    bank: str
    alias: str
    tail: str
    balance: int


class AccountDetailData(BaseModel):
    """GET /api/v1/ui/account/{account_id}"""

    account: AccountDetailInfo
    recent: list[RecentTxItem]


# --- spending (소비 분석, GET /api/v1/ui/spending) ---


class PieDatum(BaseModel):
    name: str
    value: int = Field(description="비중(%)")
    color: str
    amount: int


class ChangeItem(BaseModel):
    name: str
    amount: int


class BarCatDatum(BaseModel):
    name: str
    change: int = Field(description="전월 대비 증감률(%)")
    prev: int
    curr: int
    added: list[ChangeItem]
    removed: list[ChangeItem]


class MonthlySpendDatum(BaseModel):
    month: str
    amount: int


class CatTxDatum(BaseModel):
    name: str
    date: str
    amount: int


class SpendingData(BaseModel):
    """GET /api/v1/ui/spending"""

    pie: list[PieDatum]
    bar: list[BarCatDatum]
    monthly: list[MonthlySpendDatum]
    catTx: dict[str, list[CatTxDatum]]  # noqa: N815


# --- transactions (거래 내역, GET /api/v1/ui/transactions) ---
class TransactionItem(BaseModel):
    id: int
    name: str
    emoji: str
    date: str
    month: str
    day: int
    amount: int
    type: str = Field(description="'in' | 'out'")
    category: str


class TransactionsData(BaseModel):
    """GET /api/v1/ui/transactions"""

    months: list[str]
    items: list[TransactionItem]


# --- budget (예산 현황, GET /api/v1/ui/budget) ---


class BudgetItem(BaseModel):
    cat: str
    used: int
    total: int


class SubscriptionItem(BaseModel):
    name: str
    amount: int
    active: bool


class BudgetData(BaseModel):
    """GET /api/v1/ui/budget"""

    budgetItems: list[BudgetItem]  # noqa: N815
    subItems: list[SubscriptionItem]  # noqa: N815


# --- cards (카드 관리, GET /api/v1/ui/cards) ---


class CreditCard(BaseModel):
    name: str
    num: str
    exp: str
    bg: str


class CardsData(BaseModel):
    """GET /api/v1/ui/cards"""

    cards: list[CreditCard]

File: frontend/src/shared/types/ui.ts
// 백엔드 backend/schemas/ui.py (UI Data API, BFF) 와 1:1 대응.
// component SSE 시그널(ADR-002) 이후 FE 가 이 형태를 fetch 한다.

export interface AccountSummary {
  id: number;
  bank: string;
  alias: string;
  tail: string;
  balance: number;
  color: string;
}

export interface BalanceData {
  total: number;
  accounts: AccountSummary[];
}

// --- spending (소비 분석, GET /ui/spending) ---

export interface PieDatum {
  name: string;
  value: number;
  color: string;
  amount: number;
}

export interface ChangeItem {
  name: string;
  amount: number;
}

export interface BarCatDatum {
  name: string;
  change: number;
  prev: number;
  curr: number;
  added: ChangeItem[];
  removed: ChangeItem[];
}

export interface MonthlySpendDatum {
  month: string;
  amount: number;
}

export interface CatTxDatum {
  name: string;
  date: string;
  amount: number;
}

export interface SpendingData {
  pie: PieDatum[];
  bar: BarCatDatum[];
  monthly: MonthlySpendDatum[];
  catTx: Record<string, CatTxDatum[]>;
}

// --- transactions (거래 내역, GET /ui/transactions) ---

export interface TransactionItem {
  id: number;
  name: string;
  emoji: string;
  date: string;
  month: string;
  day: number;
  amount: number;
  type: 'in' | 'out';
  category: string;
}

export interface TransactionsData {
  months: string[];
  items: TransactionItem[];
}

// --- budget (예산 현황, GET /ui/budget) ---

export interface BudgetItem {
  cat: string;
  used: number;
  total: number;
}

export interface SubscriptionItem {
  name: string;
  amount: number;
  active: boolean;
}

export interface BudgetData {
  budgetItems: BudgetItem[];
  subItems: SubscriptionItem[];
}

// --- cards (카드 관리, GET /ui/cards) ---

export interface CreditCard {
  name: string;
  num: string;
  exp: string;
  bg: string;
}

export interface CardsData {
  cards: CreditCard[];
}
